File: 328-odd-even-linked-list/odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        
        even_head = head.next
        last_odd_node = head

        is_odd = True
        prev_node = even_head
        current_node = even_head.next if even_head else None

        while current_node:

            next_node = current_node.next

            if is_odd:
                # Move the node

                # Remove from the list
                prev_node.next = current_node.next

                # Put into the odd group
                last_odd_node.next = current_node
                last_odd_node = current_node
                last_odd_node.next = even_head
            else:
                prev_node = current_node
            

            # Move to the next
            is_odd = not is_odd
            current_node = next_node

        return head

File: 328-odd-even-linked-list/odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        
        if (head == null) {
            return null;
        }
        
        ListNode lastOddNode = head;
        ListNode evenHead = head.next;
        
        var currentNode = evenHead == null ? null : evenHead.next;
        var prevNode = evenHead;
        var isOdd = true;
        
        while (currentNode != null) {
            
            final var nextNode = currentNode.next;
            
            if (isOdd) {
                // Remove from the list
                prevNode.next = nextNode;
                
                // Put into the odd group
                lastOddNode.next = currentNode;
                lastOddNode = currentNode;
                
                // Append the even group
                lastOddNode.next = evenHead;
            } else {
                prevNode = currentNode;
            }
            
            // Move to the next node
            currentNode = nextNode;
            isOdd = !isOdd;
        }
        
        return head;
    }
}